File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/associations/JoinTest.scala
package com.github.debop4s.data.tests.mapping.associations

import com.github.debop4s.core.ValueObject
import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.model.HibernateEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.util
import java.util.Date
import javax.persistence._
import org.hibernate.annotations.{Cascade, FetchMode, Fetch, GenerationTime}
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.jpa.repository.JpaRepository


/**
 * JoinTest 
 *
 * @author sunghyouk.bae@gmail.com
 * @since 2014. 2. 28.
 */
class JoinTest extends AbstractJpaTest {

    @PersistenceContext val em: EntityManager = null
    @Autowired val userRepository: JoinUserRepository = null
    @Autowired val customerRepository: JoinCustomerRepository = null

    @Test
    def configurationTest() {
        assert(emf != null)
        assert(em != null)
        assert(userRepository != null)
        assert(customerRepository != null)
    }
}

@Embeddable
class JoinAddress extends ValueObject {

    var street: String = _
    var city: String = _
    var zipcode: String = _

    override def hashCode(): Int = Hashs.compute(street, city, zipcode)
}

@Entity
class JoinAddressEntity extends HibernateEntity[Long] {

    @Id
    @GeneratedValue
    val id: Long = 0

    var street: String = _
    var city: String = _
    var zipcode: String = _

    override def getId: Long = id

    override def hashCode(): Int = Hashs.compute(street, city, zipcode)
}

@Entity
@Table(name = "JoinCustomer")
@SecondaryTable(name = "JoinCustomerAddress", pkJoinColumns = Array(new PrimaryKeyJoinColumn(name = "customerId")))
class JoinCustomer extends HibernateEntity[java.lang.Long] {

    @Id
    @GeneratedValue
    val id: Long = 0

    override def getId: java.lang.Long = 0

    var name: String = _
    var email: String = _

    @Embedded
    @AttributeOverrides(Array(new AttributeOverride(name = "street", column = new Column(name = "street", table = "JoinCustomerAddress")),
                                 new AttributeOverride(name = "city", column = new Column(name = "city", table = "JoinCustomerAddress")),
                                 new AttributeOverride(name = "zipcode", column = new Column(name = "zipcode", table = "JoinCustomerAddress"))
                             ))
    val joinAddress: JoinAddress = new JoinAddress()

    @Temporal(TemporalType.TIMESTAMP)
    @org.hibernate.annotations.Generated(GenerationTime.INSERT)
    @Column(insertable = false, updatable = false)
    var createdAt: Date = _

    @Temporal(TemporalType.TIMESTAMP)
    @org.hibernate.annotations.Generated(GenerationTime.ALWAYS)
    @Column(insertable = false, updatable = false)
    var updatedAt: Date = _

    override def hashCode(): Int = Hashs.compute(name, email)
}


@Entity
class JoinUser extends HibernateEntity[java.lang.Long] {

    @Id
    @GeneratedValue
    val id: java.lang.Long = 0

    override def getId: java.lang.Long = 0

    var name: String = _

    // @OneToMany 를 이용한 Mapping 대상은 Entity여야 합니다.
    //
    @OneToMany(cascade = Array(CascadeType.ALL), orphanRemoval = true)
    @JoinTable(name = "JoinUserAddressMap")
    @MapKeyColumn(name = "nick")
    @ElementCollection(targetClass = classOf[JoinAddressEntity], fetch = FetchType.EAGER)
    @Fetch(FetchMode.SUBSELECT)
    var addresses: util.Map[String, JoinAddressEntity] = new util.HashMap[String, JoinAddressEntity]()

    @JoinTable(name = "JoinUserNicknameMap", joinColumns = Array(new JoinColumn(name = "userId")))
    @ElementCollection(targetClass = classOf[String], fetch = FetchType.EAGER)
    @Cascade(value = Array(org.hibernate.annotations.CascadeType.ALL))
    var nicknames: util.Set[String] = new util.HashSet[String]()

    override def hashCode(): Int = Hashs.compute(name)
}

// Repository는 Java로 만들어야겠다. QueryDSL이 제재로 안되네...

trait JoinCustomerRepository
    extends JpaRepository[JoinCustomer, java.lang.Long] {
    //with QueryDslPredicateExecutor[JoinCustomer] {

    def findByName(name: String): JoinCustomer

    def findByNameLike(name: String): util.List[JoinCustomer]

    def findByEmail(email: String): JoinCustomer
}

trait JoinUserRepository
    extends JpaRepository[JoinUser, java.lang.Long] {
    // with QueryDslPredicateExecutor[JoinUser] {

}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/associations/join/JoinTest.java
package debop4s.data.orm.jtests.mapping.associations.join;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import debop4s.data.orm.jtests.mapping.associations.join.repository.JoinCustomerRepository;
import debop4s.data.orm.jtests.mapping.associations.join.repository.JoinUserRepository;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;


@Slf4j
@Transactional
public class JoinTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Autowired JoinUserRepository userRepository;

    @Autowired JoinCustomerRepository customerRepository;

    @Test
    public void configurationTest() {
        assertThat(userRepository).isNotNull();
    }

    @Test
    public void joinUserTest() {
        JoinUser user = new JoinUser();
        user.getNicknames().add("debop");
        user.getNicknames().add("sunghyouk");

        JoinAddressEntity home = new JoinAddressEntity();
        home.setCity("Seoul");
        home.setStreet("Jungreung");
        home.setZipcode("100-100");
        user.getAddresses().put("home", home);

        JoinAddressEntity office = new JoinAddressEntity();
        office.setCity("Seoul");
        office.setStreet("Ankook");
        office.setZipcode("200-200");
        user.getAddresses().put("office", office);

        userRepository.save(user);
        em.flush();
        em.clear();

        JoinUser loaded = userRepository.findOne(user.getId());

        assertThat(loaded).isNotNull();
        assertThat(loaded.getAddresses()).isNotNull();
        assertThat(loaded.getAddresses().size()).isEqualTo(2);
        assertThat(loaded.getNicknames().size()).isEqualTo(2);

        userRepository.delete(loaded);
        userRepository.flush();

        loaded = userRepository.findOne(user.getId());
        assertThat(loaded).isNull();
    }

    @Test
    public void joinCustomerTest() {

        JoinCustomer customer = new JoinCustomer();
        customer.setName("배성혁");
        customer.setEmail("sunghyouk.bae@gmail.com");

        JoinAddress addr = new JoinAddress();
        addr.setCity("Seoul");
        addr.setStreet("Jungreung");
        addr.setZipcode("100-100");

        // Embedded Class
        customer.setJoinAddress(addr);
        customerRepository.save(customer);
        em.flush();
        em.clear();

        JoinCustomer loaded = customerRepository.findByName(customer.getName());

        assertThat(loaded).isNotNull();
        assertThat(loaded.getJoinAddress()).isNotNull();
        assertThat(loaded.getJoinAddress().getCity()).isEqualToIgnoringCase(addr.getCity());

        customerRepository.delete(loaded);

        loaded = customerRepository.findByName(customer.getName());
        assertThat(loaded).isNull();
    }

}
